draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+messages for further processing.
+
+.. admonition:: Example
+
+  We'd like to design a feature which locks Github issues immediately after
+  they are closed.
+
+  Github uses Webhooks to notify us when an issue is closed.
+
+  We set up a Data Source which starts an HTTPS server and expects incoming
+  HTTP requests on an endpoint.
+
+  Whenever a request arrives a :term:`Document Message` is published.
+
+Data Sinks
+~~~~~~~~~~
+
+A result from a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+produces a :term:`Document Message` which a Data Sink or multiple Data Sinks
+consume.
+
+These :term:`Document Messages <Document Message>` are then stored in the Sinks
+the task is registered to.
+
+.. admonition:: Example
+
+  We have a task which calculates the hourly average impressions of a user's
+  post over a period of time.
+
+  The BI team requires the data to be inserted to `BigQuery <https://cloud.google.com/bigquery/>`_
+  because it uses it to research the effectiveness of users posts.
+
+  However, the user-facing post analytics dashboard also requires this data
+  and the team that maintains it doesn't want to use BigQuery because it is not
+  a cost-effective solution and because they already use `MongoDB <https://mongodb.com/>`_
+  to store all user-facing analytics data.
+
+  To resolve the issue we declare that the task routes it's results to two data
+  sinks. One for the BI team and the other for the analytics team.
+
+  Each data sink is configured to insert the data to a specific table
+  or collection.
+
+Controller
+----------
+
+The Controller is responsible for managing the lifecycle of all other Celery
+components.
+
+Celery 5 is a more complex system with multiple components and will often be
+deployed in high throughput, highly available production systems.
+
+The introduction of multiple components require us to have another component
+that manages the entire Celery cluster.
+
+During the lifecycle of a worker the Controller also manages and optimizes the
+execution of tasks to ensure we maximize the utilization of all our resources
+and to prevent expected errors.
+
+.. note::
+
+  The Controller is meant to be run as a user service.
+  If the Controller is run with root privileges, a log message with
+  the warning level will be emitted.
+
+Foreman
++++++++
+
+The Foreman service is responsible for spawning the :ref:`Workers <draft/celery-5-high-level-architecture:Worker>`,
+:ref:`Routers <draft/celery-5-high-level-architecture:Router>` and
+:ref:`Schedulers <draft/celery-5-high-level-architecture:Scheduler>`.
+
+By default, the Foreman service creates sub-processes for
+all the required components. This is suitable for small scale deployments.
+
+Development Mode
+~~~~~~~~~~~~~~~~
+
+During development, if explicitly specified, the Foremen will start all of
+Celery's services in the same process.
+
+Since some of the new features in Celery require cryptographically signed
```

Comment:
Perhaps [HMAC](https://en.wikipedia.org/wiki/HMAC) signing would also be adequate in most cases as a means of verifying messages and is also much easier to manage, compared to using SSL certificates.

Reply:
How do you imagine this would work?

Reply:
A secret key (string) is introduced in the configuration and this is used to produce the signature of the payload and also verify it upon consumption. There is also an [hmac package](https://docs.python.org/3.8/library/hmac.html) included in the standard library, so implementation should be fairly easy.

The process resembles what web frameworks do with cookies, see [Django](https://code.djangoproject.com/wiki/Signing) for example and is quite common. An additional advantage is that the secret key could also be retrieved directly from Django's settings, no extra configuration needed.

I can't think of any real drawback compared to the current [SecureSerializer](https://github.com/celery/celery/blob/master/celery/security/serialization.py), apart from the fact that using SSL certificates gives the advantage of using asymmetric encryption in order to verify the signature. As far as I understand, it would be possible to allow only specific producers to have access to the private key and workers could have access only to the public key for message verification. However I am not sure if this can really work since (correct me if I am wrong) at certain cases workers must generate messages themselves. Perhaps this is a good consideration for the new architecture?